crates/fadroma-derive-canonize/src/enum.rs

```diff
@@ -0,0 +1,184 @@
+use proc_macro2::Span;
+use quote::quote;
+use syn::{
+    parse_quote, punctuated::Punctuated, token::Comma, Expr, FieldValue, Fields, Ident, ItemEnum,
+    Variant,
+};
+
+const CANONIZED_POSTFIX: &str = "Canon";
+
+pub fn generate(mut input: ItemEnum) -> proc_macro::TokenStream {
+    // Original identifier.
+    let humanized = input.ident;
+    input.ident = Ident::new(
+        &format!("{}{}", humanized, CANONIZED_POSTFIX),
+        Span::call_site(),
+    );
+
+    // Identifier for new enum.
+    let canonized = &input.ident.clone();
+    add_serde_derive(&mut input);
+
+    // Transform each variant's fields.
+    let mut variants = &mut input.variants;
+    for (_, variant) in variants.iter_mut().enumerate() {
+        if let Err(err) = transform_fields(&mut variant.fields) {
+            let err = err.into_compile_error();
+
+            return proc_macro::TokenStream::from(quote!(#err));
+        }
+    }
+
+    let canonized_impls = generate_match_arms(&mut variants, &humanized, &canonized, true);
+    let humanized_impls = generate_match_arms(&mut variants, &canonized, &humanized, false);
+
+    proc_macro::TokenStream::from(quote! {
+        #input
+        impl fadroma::prelude::Canonize for #humanized {
+            type Output = #canonized;
+
+            fn canonize(self, api: &impl cosmwasm_std::Api) -> cosmwasm_std::StdResult<Self::Output> {
+                Ok(match self {
+                    #canonized_impls
+                })
+            }
+        }
+        impl fadroma::prelude::Humanize for #canonized {
+            type Output = #humanized;
+
+            fn humanize(self, api: &impl cosmwasm_std::Api) -> cosmwasm_std::StdResult<Self::Output> {
+                Ok(match self {
+                    #humanized_impls
+                })
+            }
+        }
+    })
+}
+
+fn add_serde_derive(strukt: &mut ItemEnum) {
```

Comment:
This too: https://github.com/hackbg/fadroma/blob/unstable/crates/fadroma-derive-canonize/src/non_generic.rs#L49-L53
Both `ItemStruct` and `ItemEnum` have the same `Vec<Attribute>` field so it should be possible to change the signature to accept that.